TeamCode/src/main/java/org/firstinspires/ftc/teamcode/MecanumTeleOp.java

```diff
@@ -1,43 +1,99 @@
 
         waitForStart();
         if (isStopRequested()) return;
 
         while (opModeIsActive()) {
-            double y = -gamepad1.left_stick_y;
-            double x = gamepad1.left_stick_x * 1.1;
+
+            Orientation angles = gyro.getAngularOrientation(AxesReference.INTRINSIC, AxesOrder.ZYX, AngleUnit.RADIANS);
+            double botheading = angles.firstAngle;
+            telemetry.addLine()
+                    .addData("heading", formatAngle(angles.angleUnit, angles.firstAngle))
+                    .addData("roll", formatAngle(angles.angleUnit, angles.secondAngle))
+                    .addData("pitch", "%s deg", formatAngle(angles.angleUnit, angles.thirdAngle))
+                    .addData("botheading",formatAngle(angles.angleUnit,botheading));
+
+            telemetry.update();
+            double y = gamepad1.left_stick_y; // Remember, Y stick value is reversed
+            double x = gamepad1.left_stick_x * 1.1; // Counteract imperfect strafing
             double rx = gamepad1.right_stick_x;
 
+            double rotX = x * Math.cos(-botheading) + y * Math.sin(-botheading);
+            double rotY = x * Math.sin(-botheading) - y * Math.cos(-botheading);
+            telemetry.addLine()
+                    .addData("rotX", rotX)
+                    .addData("rotY", rotY);
+            telemetry.update();
             // Denominator is the largest motor power (absolute value) or 1
             // This ensures all the powers maintain the same ratio,
             // but only if at least one is out of the range [-1, 1]
-            double denominator = Math.max(Math.abs(y) + Math.abs(x) + Math.abs(rx), 1);
-            double frontLeftPower = (y + x + rx) / denominator;
-            double backLeftPower = (y - x + rx) / denominator;
-            double frontRightPower = (y - x - rx) / denominator;
-            double backRightPower = (y + x - rx) / denominator;
-
-            frontLeft.setPower(frontLeftPower);
-            backLeft.setPower(backLeftPower);
-            frontRight.setPower(frontRightPower);
-            backRight.setPower(backRightPower);
+            double denominator = Math.max(Math.abs(rotY) + Math.abs(rotX) + Math.abs(rx), 1);
+            double frontLeftPower = (rotY + rotX + rx) / denominator;
+            double backLeftPower = (rotY - rotX + rx) / denominator;
+            double frontRightPower = (rotY - rotX - rx) / denominator;
+            double backRightPower = (rotY + rotX - rx) / denominator;
+
+            hardware.frontLeft.setPower(frontLeftPower/2);
+            hardware.backLeft.setPower(backLeftPower/2);
+            hardware.frontRight.setPower(frontRightPower/2);
+            hardware.backRight.setPower(backRightPower/2);
+
+            telemetry.addData("fl power", frontLeftPower);
+            telemetry.addData("fr power", frontRightPower);
+            telemetry.addData("bl power", backLeftPower);
+            telemetry.addData("br power", backRightPower);
```

Comment:
```suggestion
            telemetry.addData("br power", backRightPower);
            telemetry.update();
```
Only update telemetry once per loop, or you lose the rest of the data.